checkpoint/mysql/mysql.go

```diff
@@ -0,0 +1,158 @@
+package mysql
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"sync"
+	"time" // this is the mysql package so it makes sense to be here
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+type key struct {
+	streamName string
+	shardID    string
+}
+
+// Option is used to override defaults when creating a new Checkpoint
+type Option func(*Checkpoint)
+
+// WithMaxInterval sets the flush interval
+func WithMaxInterval(maxInterval time.Duration) Option {
+	return func(c *Checkpoint) {
+		c.maxInterval = maxInterval
+	}
+}
+
+// Checkpoint stores and retrieves the last evaluated key from a DDB scan
+type Checkpoint struct {
+	appName     string
+	tableName   string
+	conn        *sql.DB
+	mu          *sync.Mutex // protects the checkpoints
+	done        chan struct{}
+	checkpoints map[key]string
+	maxInterval time.Duration
+}
+
+// New returns a checkpoint that uses Mysql for underlying storage
+// Using connectionStr turn it more flexible to use specific db configs
+func New(appName, tableName, connectionStr string, opts ...Option) (*Checkpoint, error) {
+	if appName == "" {
+		return nil, errors.New("application name not defined")
+	}
+
+	if tableName == "" {
+		return nil, errors.New("table name not defined")
+	}
+
+	conn, err := sql.Open("mysql", connectionStr)
```

Comment:
if we passed in the `conn` to `New` would the Postgres and Mysql implementation be interchangeable? Nothing else jumps out to me as different from the two checking point implementations

Reply:
Yeah, the duplicating everything does have a smell to it. The only other thing that changed is the upsert query in the save function. The select function changed a bit too, just `$1` syntax to `?` for params.
For mysql:
```go
upsertCheckpoint := fmt.Sprintf(`REPLACE INTO %s (namespace, shard_id, sequence_number) VALUES (?, ?, ?)`, c.tableName)
```
vs for PG:
```go
upsertCheckpoint := fmt.Sprintf(`INSERT INTO %s (namespace, shard_id, sequence_number)
				VALUES($1, $2, $3)
				ON CONFLICT (namespace, shard_id)
				DO 
				UPDATE 
				SET sequence_number= $3;`, c.tableName)
```

Could definitely refactor to have a more generic sql module that msql and pg modules are based off of where you just provide the conn string and upsert statement. I felt that was a bit invasive for my first PR on this though 😛

Reply:
got it. ok sounds good. i'm on board w/ keeping the duplication for the time being.

there might also be a world where we have templates for each of the different dbs 

```go
var upsertTemplates = map[string]string{
	"postgres": `
		INSERT INTO %s (namespace, shard_id, sequence_number)
		VALUES($1, $2, $3)
		ON CONFLICT (namespace, shard_id)
		DO UPDATE SET sequence_number= $3;
	`,
	"mysql": `
		REPLACE INTO %s (namespace, shard_id, sequence_number)
		VALUES (?, ?, ?)
	`,
}
```